README.md
# Contentful Mirror

Contentful Module for the [Magic Mirror](https://github.com/MichMich/MagicMirror). It provides a view environment for Contentful entries.

## Contentful
[Contentful](http://www.contentful.com) is a content management platform for web applications,
mobile apps and connected devices. It allows you to create, edit & manage content in the cloud
and publish it anywhere via powerful API. Contentful offers tools for managing editorial
teams and enabling cooperation between organizations.

## Installation
1. Navigate into your MagicMirror's `modules` folder and execute: `git clone https://github.com/contentful/contentful_mirror`. A new folder will appear, navigate into it.
2. Execute `npm install` to install the dependencies.

## Using the module

To use this module, add it to the modules array in the `config/config.js` file:
````javascript
modules: [
  {
    module: 'contentful_mirror',
    position: 'top_center',  // This can be any of the regions. Best results in top_center.
    config: {
      // See 'Configuration options' for more information.
    }
  }
]
````

## Configuration options

The following properties can be configured:


<table width="100%">
  <!-- why, markdown... -->
  <thead>
    <tr>
      <th>Option</th>
      <th width="100%">Description</th>
    </tr>
  <thead>
  <tbody>
    <tr>
      <td><code>maximumContentLength</code></td>
      <td>Maximum Length for a field. Truncates after this point.<br>
        <br><b>Possible values:</b> any <code>int</code></code>
        <br><b>Default value:</b> <code>1000</code>
      </td>
    </tr>
    <tr>
      <td><code>updateInterval</code></td>
      <td>Speed of change between entries.<br>
        <br><b>Possible values:</b> any <code>int</code> (in miliseconds)
        <br><b>Default value:</b> <code>10000</code>
      </td>
    </tr>
    <tr>
      <td><code>fadeSpeed</code></td>
      <td>Speed of animation.<br>
        <br><b>Possible values:</b> any <code>int</code> (in miliseconds)
        <br><b>Default value:</b> <code>4000</code>
      </td>
    </tr>
    <tr>
      <td><code>displayText</code></td>
      <td>Text displayed while loading.<br>
        <br><b>Possible values:</b> any <code>string</code>
        <br><b>Default value:</b> <code>Loading...</code>
      </td>
    </tr>
    <tr>
      <td><code>contentful</code></td>
      <td>Contentful configuration<br>
        <br><b>Possible values:</b> <code>object</code>
        <br><b>Default value:</b> <code>Described below</code>
      </td>
    </tr>
  </tbody>
</table>

### Contentful Configuration Section

* Defaults:

```js
  contentful: {
    space: "cfexampleapi",
    accessToken: "b4c0n73n7fu1",
    preview: false,
    query: {
      'content_type': 'cat'
    },
    template: "<h4>{{name}}</h4>\n" +
              "<img src='{{image.fields.file.url}}' />\n" +
              "<p class='xsmall normal'>Has {{lives}} lives.</p>\n"
  }
```

<table width="100%">
  <!-- why, markdown... -->
  <thead>
    <tr>
      <th>Option</th>
      <th width="100%">Description</th>
    </tr>
  <thead>
  <tbody>
    <tr>
      <td><code>space</code></td>
      <td>Contentful Space ID.<br>
        <br><b>Possible values:</b> any <code>string</code></code>
        <br><b>Default value:</b> <code>cfexampleapi</code>
      </td>
    </tr>
    <tr>
      <td><code>accessToken</code></td>
      <td>Production (or Preview) Access Token. Use Preview Token if <code>preview</code> is enabled<br>
        <br><b>Possible values:</b> any <code>string</code>
        <br><b>Default value:</b> <code>b4c0n73n7fu1</code>
      </td>
    </tr>
    <tr>
      <td><code>preview</code></td>
      <td>Enables Preview API.<br>
        <br><b>Possible values:</b> <code>boolean</code>
        <br><b>Default value:</b> <code>false</code>
      </td>
    </tr>
    <tr>
      <td><code>query</code></td>
      <td>Contentful CDA Query for Entries endpoint. Content Type is advised to be always sent for templates to work.<br>
        <br><b>Possible values:</b> any <code>object</code>
        <br><b>Default value:</b> <code>{'content_type': 'cat'}</code>
      </td>
    </tr>
    <tr>
      <td><code>template</code></td>
      <td>HTML Template to display content in the mirror.<br>
        <br><b>Possible values:</b> any <code>string</code>
        <br><b>Default value:</b> <code>Described below</code>
      </td>
    </tr>
  </tbody>
</table>

### Templates

This module uses a very simple template engine. It basically fetches anything between a pair of braces e.g. `{{foo}}`, and sends it as a message to the `entry.fields` object.
You can chain as many messages as you want, and the engine will send them to the result.

For example:

If I have the following entry

```json
{
  "fields": {
    "foo": "bar",
    "baz": {
      "fields": {
        "nyan": "cat"
      }
    },
    "bar": [1, 2, 3]
  }
}
```

I could use inside my template `{{baz.fields.nyan}}` and it will render `cat`. Notice that past the first level of the entry, if there is a nested entry, you are required
to explicitly send the `fields` message.

The tool already does 1 level of link resolution, so if you have linked content, you can display it with this tool.

The templates can also handle array indexes, so if you could do `{{bar[1]}}` and will render `2`.

Link resolution does not work for arrays.

## Dependencies
- [contentful](https://www.npmjs.com/package/contentful) (installed by `npm install`)

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/contentful/contentful_mirror. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# CHANGELOG

## Master

### Added
* Data Fetching from Contentful - Server Side
* Include Resolution - Client Side
* Template Engine - Client Side


node_helper.js
/* Magic Mirror
 * Node Helper: Contentful Mirror
 *
 * By David Litvak Bruno (Contentful GmbH) https://contentful.com
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var contentful = require("contentful");

module.exports = NodeHelper.create({
  start: function() {
    var self = this;

    this.client = {};

    console.log("Starting node helper for: " + this.name);
  },

  createClient: function(space, accessToken, preview) {
    console.log("Creating Contentful Client for: " + space);
    this.client = contentful.createClient({
      space: space,
      accessToken: accessToken,
      preview: preview,
      resolveLinks: false
    });
  },

  getEntries: function(query) {
    var self = this;

    this.client.getEntries(query)
    .then(function(entries) {
      self.sendSocketNotification("CONTENTFUL_ENTRIES_FETCHED", {entries: entries});
    })
    .catch(function(e) {
      console.log(e);
    });
  },

  socketNotificationReceived: function(notification, payload) {
    if (notification === "CREATE_CLIENT") {
      this.createClient(payload.space, payload.accessToken);
      this.sendSocketNotification("CLIENT_CREATED");
    } else if (notification === "GET_ENTRIES") {
      this.getEntries(payload.query);
    }
  },
});


contentful_mirror.js
/* Magic Mirror
 * Module: Contentful Mirror
 *
 * By David Litvak Bruno (Contentful GmbH) https://contentful.com
 * MIT Licensed.
 */

Module.register("contentful_mirror", {
  defaults: {
    contentful: {
      space: "cfexampleapi",
      accessToken: "b4c0n73n7fu1",
      preview: false,
      query: {
        'content_type': 'cat'
      },
      template: "<h4>{{name}}</h4>\n" +
                "<img src='{{image.fields.file.url}}' />\n" +
                "<p class='xsmall normal'>Has {{lives}} lives.</p>\n"
    },
    maximumContentLength: 1000,
    updateInterval: 10000,
    fadeSpeed: 4000,
    displayText: "Loading..."
  },

  start: function() {
    Log.info("Starting module: " + this.name);

    this.displayText = this.config.displayText;

    this.sendSocketNotification("CREATE_CLIENT", {
      space: this.config.contentful.space,
      accessToken: this.config.contentful.accessToken,
      preview: this.config.contentful.preview
    });

    var self = this;
    setInterval(function() {
      self.updateDom(self.config.fadeSpeed);
    }, this.config.updateInterval);
  },

  socketNotificationReceived: function(notification, payload) {
    if (notification === "CLIENT_CREATED") {
      this.sendSocketNotification("GET_ENTRIES", {query: this.config.contentful.query});
    } else if (notification === "CONTENTFUL_ENTRIES_FETCHED") {
      if (payload.entries.items.length > 0) {
        this.entries = payload.entries;
        this.resolveIncludes();
      }
    } else if (notification === "FETCH_ERROR") {
      this.displayText = "Error fetching data";
      Log.error("Contentful Error. Could not fetch data.");
    } else {
      Log.log("Contentful Mirror received an unknown socket notification: " + notification);
    }

    this.updateDom(this.config.fadeSpeed);
  },

  resolveIncludes: function() {
    var includes = this.entries.includes;
    includes.Entry = includes.Entry || [];
    includes.Entry = includes.Entry.concat(this.entries.items).filter(function(value, index, self) {
      return self.indexOf(value) === index;
    });

    this.entries.items.forEach(function(entry) {
      Object.keys(entry.fields).forEach(function(key) {
        if (typeof entry.fields[key] === "object" &&
              entry.fields[key] !== null &&
              entry.fields[key].sys !== undefined) {
          var included = includes[entry.fields[key].sys.linkType].find(function(include) {
            return include.sys.id === entry.fields[key].sys.id;
          });
          entry.fields[key] = included;
        }
      });
    });
  },

  getRandomEntry: function() {
    return this.entries.items[Math.floor(Math.random() * this.entries.items.length)];
  },

  getField: function(entry, field_accessor) {
    var accessors = field_accessor.split(".");
    var field = null;
    var index_regex = /\[(\d+)\]/g;

    accessors.forEach(function (e) {
      var index = e.match(index_regex);
      if (index !== null) {
        index = index[1];
      }
      e = e.replace(index_regex, "");

      if (field === null) {
        field = entry.fields[e];
      } else if (field === undefined) {
        return field;
      } else {
        field = field[e];
      }

      if (typeof field === "array" && index !== null) {
        field[index];
      }
    }, this);

    return field;
  },

  getContentfulRepresentation: function() {
    if (this.entries !== undefined) {
      // We fetch a random entry
      var selected_entry = this.getRandomEntry();
      var template_regex = /\{\{\s*(?:\w+(?:\[\d+\])?\.?)+\s*\}\}/gm;

      var self = this;
      return this.config.contentful.template.replace(template_regex, function(match) {
        var field = self.getField(selected_entry, match.replace("{{", "").replace("}}", ""));

        if (typeof field === "string") {
          if (field.length > self.config.maximumContentLength) {
            field = field.substring(0, self.config.maximumContentLength) + "...";
          }

          if (field.search("\n")) {
            field = field.replace(/\n/g, "<br>");
          }
        }

        if (field === null || field === undefined) {
          field = "";
        }

        return field;
      });
    } else {
      return this.displayText;
    }
  },

  getDom: function() {
    var content =this.getContentfulRepresentation();
    var wrapper = document.createElement("div");
    wrapper.className = "contentful thin medium bright";
    wrapper.innerHTML = content;

    return wrapper;
  }
});
